packages/front-end/components/Experiment/EditDOMMutationsModal.tsx

```diff
@@ -40,7 +47,100 @@ const EditDOMMutatonsModal: FC<{
     [newVisualChange, setNewVisualChange]
   );
 
+  const setDOMMutation = useCallback(
+    (index: number, updates) => {
+      setNewVisualChange({
+        ...newVisualChange,
+        domMutations: newVisualChange.domMutations.map((m, i) =>
+          i === index ? updates : m
+        ),
+      });
+    },
+    [newVisualChange, setNewVisualChange]
+  );
+
+  const setDOMMutationStr = useCallback(
+    (index: number, str: string) => {
+      setNewDOMMutationStr((strs) => {
+        const newStrs = [...strs];
+        newStrs[index] = str;
+        return newStrs;
+      });
+    },
+    [setNewDOMMutationStr]
+  );
+
+  const setDOMMutationErrors = useCallback(
+    (index: number, error: string) => {
+      setNewDOMMutationErrors((errors) => {
+        const newErrors = [...errors];
+        newErrors[index] = error;
+        return newErrors;
+      });
+    },
+    [setNewDOMMutationErrors]
+  );
+
+  const validateDOMMutations = useCallback(
+    (index: number, mutation: string) => {
+      try {
+        const m = JSON.parse(mutation);
+        /*
+        valid DOM mutation object. No other keys are allowed:
+          selector: string;
+          attribute: string;
+          action: 'append' | 'set' | 'remove';
+          value?: string;
+          parentSelector?: string;
+          insertBeforeSelector?: string;
+         */
+        if (m.selector === undefined) {
+          throw new Error("selector key is required");
+        }
+        if (m.attribute === undefined) {
+          throw new Error("attribute key is required");
+        }
+        if (m.action === undefined) {
+          throw new Error("action key is required");
+        }
+        if (!["append", "set", "remove"].includes(m.action)) {
+          throw new Error("action must be one of 'append', 'set', or 'remove'");
+        }
+        // check to make sure the object has no non-defined keys
+        if (Object.keys(m).length > 3) {
+          Object.keys(m).forEach((key) => {
+            if (
+              ![
+                "selector",
+                "attribute",
+                "action",
+                "value",
+                "parentSelector",
+                "insertBeforeSelector",
+              ].includes(key)
+            ) {
+              throw new Error(`Invalid key: ${key}`);
+            }
+          });
+        }
+
+        setDOMMutation(index, m);
+        setDOMMutationErrors(index, "");
+      } catch (e) {
+        setDOMMutationErrors(index, e.message);
+      }
+    },
+    [setDOMMutation, setDOMMutationErrors]
+  );
+
   const onSubmit = () => {
+    // make sure all DOM mutations are valid
+    newDOMMutationStr.forEach((m, i) => {
+      validateDOMMutations(i, m);
+    });
+    if (newDOMMutationErrors.some((e) => e)) {
```

Comment:
This won't work as expected.  The `setDOMMutationErrors` call within `validateDOMMutations` doesn't take effect until the next render.  This will likely still work because you're also calling `validateDOMMutations` when the input changes, but it is a potential race condition.  Better to have the validate method return true or false and check the return value here.